test/src/org/apache/jmeter/protocol/http/sampler/TestHTTPSamplers.java

```diff
@@ -364,4 +369,23 @@ public void testSetAndGetMimetype() {
         file = sampler.getHTTPFiles()[0];
         assertEquals("mime2", file.getMimeType());
     }
+
+    @Test
+    public void testRawBodyFromFile() {
+        String baseDirPath = FileServer.getFileServer().getBaseDir();
+        File baseDir = new File(baseDirPath);
+        try {
+            FileServer.getFileServer().setBase(baseDir.getParentFile());
+            HTTPSamplerBase sampler = new HTTPSampler3();
+            sampler.setMethod("POST");
+            sampler.setPath("http://httpbin.org/post");
+            sampler.setHTTPFiles(new HTTPFileArg[]{new HTTPFileArg("bin/jmeter.properties", "", "")});
+
+            SampleResult sample = sampler.sample();
+            System.out.println(sample.getResponseDataAsString());
+            assertFalse(sample.getResponseDataAsString().contains("java.io.FileNotFoundException:"));
```

Comment:
@artem-fedorov , @FSchumacher , please refrain from computations inside asserttrue/false.
It produces just a failure like

```
java.lang.AssertionError
        at org.junit.Assert.fail(Assert.java:86)
        at org.junit.Assert.assertTrue(Assert.java:41)
        at org.junit.Assert.assertFalse(Assert.java:64)
        at org.junit.Assert.assertFalse(Assert.java:74)
        at org.apache.jmeter.protocol.http.sampler.TestHTTPSamplers.testRawBodyFromFile(TestHTTPSamplers.java:385)
```

and it provides no clue on what really went wrong.

It would be so much better if you used

```
String msg = sample.getResponseDataAsString();
assertFalse("sample.getResponseDataAsString()"+
    " should not contain java.io.FileNotFoundException,"+
    " but the actual value is " + msg);
```
Do you see? Thank you for the understanding.

Reply:
@vlsi I'm agree with your remark. The second code style is better for understanding in fail cases.

Reply:
Thats is definitely better, but I think it would be even better to use hamcrests `StringContains.constainsString` and junits `Assert.assertThat` for the test.

Using static imports it would result in
  ```java
assertThat(sample.getResponseDataAsString(), not(containsString("java.io.FileNotFoundException:")));
````

Reply:
Mr @pmouawad  might argue that Groovy+Spock might even be better (e.g. https://twitter.com/VladimirSitnikv/status/1100666414548045825 ).

However please estimate how assert failures would look like. Would it look like a good bug report? Is the failure message+stacktrace enough to file a understandable bug report?

What we typically see is that "tests pass". We don't accept code changes if tests fail, do we?

However the important bit of the test code is the way it fails should something wrong happen. It is important that developers understand that, and committers acting as gatekeepers should watch for that as well.

I don't mean "assert messages" are the top priority, however having sane messages is vital for further maintenance of the codebase.

Reply:
> Mr @pmouawad might argue that Groovy+Spock might even be better (e.g. https://twitter.com/VladimirSitnikv/status/1100666414548045825 ).

I don't know if it's sarcastic or not :-)
Still, I find the messages very clear and text comparison a great improvement. 
Integration with Eclipse is not that great (at least from my experience) but still interesting framework (Thanks Graham Russell (to Caesar what's belong to Caesar) for making me discover it). No firm conviction on JUnit 5 or Spock still, both do the job

Reply:
I like the spock error messages, too. They might be a bit over the top and frightening to the beginner, but very descriptive and therefore helpful.

> ... Would it look like a good bug report? Is the failure message+stacktrace enough to file a understandable bug report?

The error message would look like:
```
Expected: not a string containing "java.io.FileNotFoundException"
  but was: "the real content of the sampler ..."
```

In my opinion this is a good (enough) message, with the plus side, that is is code only and thus better suited for refactoring. But either way, we should get rid of simple `assertFalse` or `assertTrue` statements, that have no meaningful description message.

> What we typically see is that "tests pass". We don't accept code changes if tests fail, do we?

What do you mean by this? If the code is wrong, or the error messages are difficult to read, we should -- and have done -- change the tests.

All in all, this seems to grow into a discussion, that would be good to have on the mailing list.